src/dipdup/http.py

```diff
@@ -0,0 +1,120 @@
+import asyncio
+import hashlib
+import logging
+import pickle
+import platform
+from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+
+import aiohttp
+from aiolimiter import AsyncLimiter
+from fcache.cache import FileCache  # type: ignore
+
+from dipdup import __version__
+from dipdup.config import HTTPConfig  # type: ignore
+
+
+class HTTPGateway(ABC):
+    def __init__(self, url: str, http_config: Optional[HTTPConfig] = None) -> None:
+        self._http_config = self._default_http_config()
+        self._http_config.merge(http_config)
+        self._http = _HTTPGateway(url.rstrip('/'), self._http_config)
+
+    @abstractmethod
+    def _default_http_config(self) -> HTTPConfig:
+        ...
+
+    async def close_session(self) -> None:
+        await self._http.close_session()
+
+    def set_user_agent(self, *args: str) -> None:
+        self._http.set_user_agent(*args)
+
+
+class _HTTPGateway:
+    """Wrapper for aiohttp HTTP requests.
+
+    Covers caching, retrying failed requests and ratelimiting"""
+
+    def __init__(self, url: str, config: HTTPConfig) -> None:
+        self._logger = logging.getLogger('dipdup.http')
+        self._url = url
+        self._config = config
+        self._user_agent_args: Tuple[str, ...] = ()
+        self._user_agent: Optional[str] = None
+        self._cache = FileCache('dipdup', flag='cs')
+        self._ratelimiter = (
+            AsyncLimiter(max_rate=config.ratelimit_rate, time_period=config.ratelimit_period)
+            if config.ratelimit_rate and config.ratelimit_period
+            else None
+        )
+        self._session = aiohttp.ClientSession()
+
+    @property
+    def user_agent(self) -> str:
+        if self._user_agent is None:
+            user_agent_args = (platform.system(), platform.machine()) + (self._user_agent_args or ())
+            user_agent = f'dipdup/{__version__} ({"; ".join(user_agent_args)})'
+            user_agent += ' ' + aiohttp.http.SERVER_SOFTWARE
+            self._user_agent = user_agent
+        return self._user_agent
+
+    async def _wrapped_request(self, method: str, url: str, **kwargs):
+        attempts = list(range(self._config.retry_count)) if self._config.retry_count else [0]
+        for attempt in attempts:
+            self._logger.debug('HTTP request attempt %s/%s', attempt + 1, self._config.retry_count)
+            try:
+                return await self._request(
+                    method=method,
+                    url=url,
+                    **kwargs,
+                )
+            except (aiohttp.ClientConnectionError, aiohttp.ClientConnectorError) as e:
```

Comment:
+ ServerDisconnectedError
+ 429

Reply:
`ServerDisconnectedError` is a subclass of `ClientConnectionError`, fixed 429.